`bevy_text::input` module with no undo

Zeophlite opened the pull request:
# Objective

- https://github.com/bevyengine/bevy/pull/20336 without undo/redo (see https://github.com/bevyengine/bevy/pull/20489 to upstream the `undo_2` library)

## Testing

- CI

alice-i-cecile commented on crates/bevy_text/src/input.rs:
```diff
@@ -0,0 +1,1312 @@
+
+/// Basic clipboard implementation that only works within the bevy app.
+#[derive(Resource, Default)]
+pub struct Clipboard(pub String);
```
I am tempted to cut this from the initial PR: we can make this simpler and improve reviewability / controversy.

alice-i-cecile replied:
Actually, the implementation looks very simple and uncontroversial. This can stay.

alice-i-cecile commented on crates/bevy_text/src/input.rs:
```diff
@@ -0,0 +1,1312 @@
+/// On changes, the text input systems will automatically update the buffer, layout and fonts as required.
+#[derive(Component, Debug, PartialEq)]
+pub struct TextInputAttributes {
+    /// The text input's font, also used for any [`Placeholder`] text or password mask.
```
```suggestion
    /// The text input's font, which also applies to any [`Placeholder`] text or password mask.
```

alice-i-cecile commented on crates/bevy_text/src/input.rs:
```diff
@@ -0,0 +1,1312 @@
+}
+
+/// If a text input entity has a `TextInputFilter` component, after each [`TextEdit`] is applied, the [`TextInputBuffer`]’s text is checked
+/// against the filter, and if it fails, the `TextEdit` is rolled back.
```
Is this immediately rolled back, or is invalid state possible? Either way this needs to be clearly documented.

alice-i-cecile reviewed (changes requested):
I have some small doc nits that I would like to resolve before we merge this, but fundamentally this is polished, well-made and uncontroversial. If we can resolve the changes and get approvals in time for 0.17 I'm comfortable shipping it, but I don't currently think we should block the release for this.

atlv24 commented on examples/2d/text_input_2d.rs:
```diff
@@ -0,0 +1,478 @@
+        #[cfg(target_os = "macos")]
+        let is_command_pressed = keyboard_state.pressed(Key::Super);
+
+        for (mut actions, mut overwrite) in query.iter_mut() {
```
is it possible to have more than one text being edited at a time? will this paste etc in all of them?

Zeophlite replied:
The original PR, https://github.com/bevyengine/bevy/pull/20326 , which `bevy_text::input` was pulled out of, gives a more complete example of multiple input boxes

Zeophlite commented on examples/2d/text_input_2d.rs:
```diff
@@ -0,0 +1,478 @@
+            let x1 = (Anchor::TOP_LEFT.0.x - anchor.as_vec().x) * size.0.x;
+            let x2 = (Anchor::TOP_LEFT.0.x - anchor.as_vec().x + 1.) * size.0.x;
+            let y1 = (Anchor::TOP_LEFT.0.y - anchor.as_vec().y - 1.) * size.0.y;
+            let y2 = (Anchor::TOP_LEFT.0.y - anchor.as_vec().y) * size.0.y;
```
I now have `let (x1, x2, y1, y2) = anchor.reposition(size.0);` but I don't think that's a good name...

viridia replied:
The name "anchor" comes from the old Mac / Xerox Parc GUI design guidelines: when dragging a selection region, the "cursor" is the moving end of the selection range (attached to the mouse cursor), and the "anchor" is the fixed part (the position at which you first clicked). However, I suspect the definition has evolved in the last 40 years.

alice-i-cecile reviewed (approved):
Good edits: I appreciate how clear your docs are.

I'm comfortable with this, and think this is an excellent first step. I'm going to float shipping this in 0.17, and if no one complains, I'm pressing the button.

atlv24 commented on release-content/release-notes/bevy_text_input_module.md:
```diff
@@ -0,0 +1,35 @@
+* Drag to select.
+* Double click to select a word.
+* Triple click to set select a line.
+* Indent and unident.
```
```suggestion
* Indent and unindent.
```

atlv24 commented on release-content/release-notes/bevy_text_input_module.md:
```diff
@@ -0,0 +1,35 @@
+* Fixes the line cropping while vertical scrolling bug in cosmic-text.
+* Text selection.
+* Cut, copy and paste.
+* Numeric input modes.
```
```suggestion
* Numeric input mode.
```

alice-i-cecile commented on release-content/migration-guides/load_font_to_fontdb_parameter_change.md:
```diff
@@ -0,0 +1,6 @@
+---
+title: `load_font_to_fontdb` parameter change
+pull_requests: [20366]
```
I think this is zombie content?

mockersf commented on examples/2d/text_input_2d.rs:
```diff
@@ -0,0 +1,475 @@
+                cursor_blink_timer: Some(0.),
+                ..Default::default()
+            },
+            TextLayoutInfo::default(),
+            TextEdits::default(),
+            TextInputAttributes::default(),
+            TextInputTarget::default(),
+            Overwrite::default(),
+            TextFont::default(),
```
all those shouldn't be needed, most should be required components

mockersf commented on examples/2d/text_input_2d.rs:
```diff
@@ -0,0 +1,475 @@
+    }
+}
+
+fn update_text_input_attributes(
```
why is this system needed? I don't think anything is changing the `TextFont` component

mockersf commented on examples/2d/text_input_2d.rs:
```diff
@@ -0,0 +1,475 @@
+    }
+}
+
+fn extract_text_input(
```
some comment on what is happening in this system would be nice

mockersf commented:
I tried using the Noto Simplified Chinese font to test IME support, but couldn't get it to be used. Placeholder text is not displayed, and text typed is using the default font.

When switching to other fonts, it works correctly. The only difference that comes to mind is that the fonts that worked are a few hundreds kb while the noto font is 17mb so it couldn't be longer to load and not available from the start

alice-i-cecile commented on examples/2d/text_input_2d.rs:
```diff
@@ -0,0 +1,453 @@
+            Transform::from_translation(Vec3::new(0., 150., 0.)),
+            Placeholder::new("type here.."),
+            Visibility::default(),
+            VisibilityClass([TypeId::of::<Sprite>()].into()),
```
This needs a comment.

alice-i-cecile commented on examples/2d/text_input_2d.rs:
```diff
@@ -0,0 +1,453 @@
+                ..Default::default()
+            },
+            TextInputAttributes::default(),
+            Overwrite::default(),
```
This should be ommitted, and simply explained in the text_input module docs.

Zeophlite replied:
Not sure what you mean here - `Overwrite` is a `bool` component in this example

alice-i-cecile replied:
Ah, my bad.

alice-i-cecile commented on examples/2d/text_input_2d.rs:
```diff
@@ -0,0 +1,453 @@
+                cursor_blink_timer: Some(0.),
+                ..Default::default()
+            },
+            TextInputAttributes::default(),
```
This should be a required component.

alice-i-cecile commented on examples/2d/text_input_2d.rs:
```diff
@@ -0,0 +1,453 @@
+            TextInputAttributes::default(),
+            Overwrite::default(),
+            TextInputSize(Vec2::new(500., 250.)),
+            Transform::from_translation(Vec3::new(0., 150., 0.)),
```
This needs a comment about how the UI transform form is needed.

alice-i-cecile commented on crates/bevy_text/src/input.rs:
```diff
@@ -0,0 +1,1369 @@
+/// This is written to in the [`apply_text_edits`] system when
+/// [`TextEdit::Copy`], [`TextEdit::Cut`] or [`TextEdit::Paste`] edits are applied.
+#[derive(Resource, Default)]
+pub struct Clipboard(pub String);
```
I would prefer to add more structure to this module to make it easier to navigate and understand. I would split this as:

- input (mod.rs)
- input/clipboard.rs
- input/placeholder.rs
- input/rendering.rs (also has layout)
- input/cursor_blink.rs
- input/filtering.rs
- input/masking.rs

Zeophlite replied:
I can do that, but was also planning on rebasing https://github.com/bevyengine/bevy/pull/20326 to top of this PR first, so am hoping to do this as a follow-up PR

alice-i-cecile commented on crates/bevy_text/src/input.rs:
```diff
@@ -0,0 +1,1369 @@
+//! * [`PasswordMask`] - Optional component that can be added to hide the text input buffer contents by replacing the characters with a mask character.
+//! * [`Placeholder`] - Optional component that can be added to display placeholder text when the input buffer is empty.
+//!
+//! The [`TextInputBuffer`] component itself also has two fields that can be configured:
```
I've documented the current behavior but I don't like it. The former should be moved into TextInputAttributes, and the second should be an optional component.

Zeophlite replied:
`space_advance` is tied to the `editor` , so I've left that in the component.  Made a new `CursorBlink` component.

alice-i-cecile reviewed (changes requested):
I've spent some time documenting this PR and making very minor tweaks. After digging in deeper, I have some complaints that I would like to resolve.

I feel very mixed about shipping this as is, without even a UI example. On the one hand, this is a very nice piece of very useful functionality. On the other hand, users will get excited about "oh wow, text input! I should use this in a UI" and promptly run into a brick wall.

viridia commented:
> I feel very mixed about shipping this as is, without even a UI example.

What needs to happen is to extract the widgets and examples from the original pr: https://github.com/bevyengine/bevy/pull/20326. This has both single line and multi line widgets, as well as multiple examples. Looking at the diffs, I think there's only a handful of large source files which would need to be copied over.

This could be done as a separate PR.

I would also make several changes:

* Migrate any changes relative to the text input code in this PR.
* Rename the text input widgets to `TextInput` and `TextArea` to be consistent with web nomenclature.
* Move the widgets into the same crate as the other core widgets (presumably after the renaming).

Once all that is done, an additional follow-on PR could implement a feathers-branded text input widget. The main downside there is that any changes made to feathers increases the amount of BSN migration technical debt.

Zeophlite commented:
Hey @tigregalis , I think the below addresses your feedback for this release, and the rest can be deferred for another PR?


> I don't understand space_advance - I assume it's working around some bug or limitation: what's the issue it's solving?

I'm not super sure either, @ickshonpe @viridia can you chime in?

> Text is always joined via a "\n" but for correctness you might use the BufferLine's LineEnding (e.g. on windows it might be \r\n if coming from a file or a paste), but I'm not 100% sure on this.

I'm not sure on this, we just call `editor.action(Action::Enter);` and `cosmic_text` handles that
 
> 20.0 is used as a magic constant in a few places, can we make it a named constant(s)?

Done

> As a default, line height of 1.2x font size is used generally across web, and looks good generally. For example I think it's in firefox's user agent stylesheet if I'm not wrong.

Added a comment to reflect this

> Comment on Paste should note it replaces the current selection, not inserting at the cursor. More broadly, all insertions replace the current selection, sometimes the selection just has a width of 0.

Added a comment

> TextEdit::Submit feels like it belongs in the realm of TextInput being a full widget, maybe even as part of a Form, so its inclusion is odd when there's not yet a full widget. It's the only variant that is special-cased in the apply_text_edits handler system. It's also special-cased in the apply_text_edit function. I would be inclined to remove it and revisit it as a higher level action with the text input widget, and not tied at all to text editing.

I agree, I've separated the "submission" into the example, this could be merged back later, depending on how "forms" look

viridia commented:
I'm not sure what `space_advance` is. However, what I can tell you is that this is an area in which desktop UIs vary.

Early terminal displays did not support I-beam style cursors, and could only blink a fixed-width character on and off; when the cursor was at the end of the line, it would blink an empty space.

In the original MacOS UI guidelines, and for many years after, the standard behavior for most GUIs was that when you selected the invisible line-break character at the end of a wrapped line, the selection highlight would be extended all the way to the right of the text input box, that is, the left edge of the rectangle would start just after the last character on the line, and the right edge of the rectangle would be the inner right padding of the containing box.

However, some modern apps (like Chrome and VSCode) do not do this: instead, selecting the wrap character produces a cursor of some fixed width. When there is no wrap character (that is, either the cursor is at the end of the text, or it broke the line in the middle of a word) it shows a normal I-beam style cursor.

tigregalis commented:
> Hey @tigregalis , I think the below addresses your feedback for this release, and the rest can be deferred for another PR?

Yeah, sounds good.

I am a little hesitant about the password mask and the input filter being part of the same PR, but if those are the features people are looking for ultimately, maybe it doesn't make sense to hold them back for the sake of code history/evolution clarity.

atlv24 commented on examples/2d/text_input_2d.rs:
```diff
@@ -0,0 +1,465 @@
+            size: size.0 * scale_factor,
+            scale_factor,
+        }) {
+            let (x1, x2, y1, y2) = anchor.reposition(size.0);
+            let new_aabb = Aabb::from_min_max(Vec3::new(x1, y1, 0.), Vec3::new(x2, y2, 0.));
```
Specifically this Aabb, and it should be called something like calc bounds or get bounds for.

atlv24 reviewed (commented):
i am concerned about how much code is in the example. additionally, it seems this is just for 2d, can it be used in a ui in 3d or etc easily?

Zeophlite commented:
> i am concerned about how much code is in the example. additionally, it seems this is just for 2d, can it be used in a ui in 3d or etc easily?

There's a lot in the example as this is the "low level" api - the follow-up PR has the "mid level" api for widgets, etc.  Have a look at the `examples/` in https://github.com/bevyengine/bevy/pull/20326/files .

Zeophlite commented:
> and some that doesn't interact well together

@mockersf Could you expand on this point?

Zeophlite commented:
I've taken the time to split up `input.rs`

Zeophlite commented:
I have also taken the time to adopt https://github.com/bevyengine/bevy/pull/20326 , rebase onto this branch - see https://github.com/bevyengine/bevy/pull/20826

This was very helpful in fleshing out some concepts, such as "submit"

rdrpenguin04 commented:
One element I would personally consider critical for an example is handling text input on a mobile browser, i.e. by popping open an on-screen keyboard when a text input is focused. This definitely feels like it belongs more in the eventual widget code, but it also would feel right to have that kind of support within Bevy (experimentally?) so work wouldn't need to be duplicated across many projects.
